engine/src/hashchain.rs

```diff
@@ -0,0 +1,889 @@
+        self.previous_block_hashchain
+    }
+
+    /// Gets the genesis block hashchain of the structure.
+    pub fn get_genesis_block_hashchain(&self) -> RawH256 {
+        self.genesis_block_hashchain
+    }
+}
+
+/// Blockchain Hashchain error module.
+pub mod blockchain_hashchain_error {
+    pub const ERR_STATE_NOT_FOUND: &[u8; 19] = b"ERR_STATE_NOT_FOUND";
+    pub const ERR_STATE_SERIALIZATION_FAILED: &[u8; 26] = b"ERR_STATE_SERIALIZE_FAILED";
+    pub const ERR_STATE_CORRUPTED: &[u8; 19] = b"ERR_STATE_CORRUPTED";
+    pub const ERR_BLOCK_HEIGHT_INCORRECT: &[u8; 26] = b"ERR_BLOCK_HEIGHT_INCORRECT";
+
+    #[derive(Debug)]
+    /// Blockchain Hashchain Error
+    pub enum BlockchainHashchainError {
+        /// The state is missing from storage, need to initialize with contract `new` method.
+        NotFound,
+        /// The state serialized had failed.
+        SerializationFailed,
+        /// The state is corrupted, possibly due to failed state migration.
+        DeserializationFailed,
+        /// The block height is incorrect regarding the current block height.
+        BlockHeightIncorrect,
+    }
+
+    impl AsRef<[u8]> for BlockchainHashchainError {
+        fn as_ref(&self) -> &[u8] {
+            match self {
+                Self::NotFound => ERR_STATE_NOT_FOUND,
+                Self::SerializationFailed => ERR_STATE_SERIALIZATION_FAILED,
+                Self::DeserializationFailed => ERR_STATE_CORRUPTED,
+                Self::BlockHeightIncorrect => ERR_BLOCK_HEIGHT_INCORRECT,
+            }
+        }
+    }
+}
+
+/// Block Hashchain Computer
+/// The order of operations should be:
+/// 1. Create the BlockHashchainComputer one time.
+/// 2. Add transactions of the current block.
+/// 3. Compute the block hashchain for the current block once all the transactions were added.
+/// 4. Clear the transactions of the current block.
+/// 5. Go back to step 2 for the next block.
+#[derive(BorshSerialize, BorshDeserialize)]
+struct BlockHashchainComputer {
+    txs_merkle_tree: StreamCompactMerkleTree,
+}
+
+impl BlockHashchainComputer {
+    pub fn new() -> Self {
+        Self {
+            txs_merkle_tree: StreamCompactMerkleTree::new(),
+        }
+    }
+
+    /// Adds a transaction.
+    pub fn add_tx(&mut self, method_name: &str, input: &[u8], output: &[u8]) {
+        let method_name_hash = keccak(method_name.as_bytes()).0;
+        let input_hash = keccak(input).0;
+        let output_hash = keccak(output).0;
+
+        let tx_hash = keccak(&[method_name_hash, input_hash, output_hash].concat()).0;
+
+        self.txs_merkle_tree.add(tx_hash);
+    }
+
+    /// Computes the block hashchain.
+    pub fn compute_block_hashchain(
+        &self,
+        chain_id_hash: RawH256,
+        contract_account_id_hash: RawH256,
+        current_block_height_hash: RawH256,
+        previous_block_hashchain: RawH256,
+    ) -> RawH256 {
+        let txs_hash = self.txs_merkle_tree.compute_hash();
+
+        keccak(
```

Comment:
This is the current hashing formula that I'm applying for a block. Do you have any comments regarding the parameters or order that I'm using?

Reply:
No comments, looks good to me.